store.js

```diff
@@ -1,174 +1,88 @@
-import {
-	assign,
-	blankObject,
-	_differs,
-	_differsImmutable,
-	get,
-	on,
-	fire
-} from './shared.js';
-
-function Store(state, options) {
-	this._handlers = {};
-	this._dependents = [];
-
-	this._computed = blankObject();
-	this._sortedComputedProperties = [];
-
-	this._state = assign({}, state);
-	this._differs = options && options.immutable ? _differsImmutable : _differs;
-}
+import { run_all } from './internal.js';
 
-assign(Store.prototype, {
-	_add(component, props) {
-		this._dependents.push({
-			component: component,
-			props: props
-		});
-	},
-
-	_init(props) {
-		const state = {};
-		for (let i = 0; i < props.length; i += 1) {
-			const prop = props[i];
-			state['$' + prop] = this._state[prop];
-		}
-		return state;
-	},
-
-	_remove(component) {
-		let i = this._dependents.length;
-		while (i--) {
-			if (this._dependents[i].component === component) {
-				this._dependents.splice(i, 1);
-				return;
-			}
-		}
-	},
+export function readable(start, value) {
+	const subscribers = [];
+	let stop;
 
-	_set(newState, changed) {
-		const previous = this._state;
-		this._state = assign(assign({}, previous), newState);
+	function set(newValue) {
+		if (newValue === value) return;
+		value = newValue;
+		subscribers.forEach(fn => fn(value));
+	}
 
-		for (let i = 0; i < this._sortedComputedProperties.length; i += 1) {
-			this._sortedComputedProperties[i].update(this._state, changed);
-		}
+	return {
+		subscribe(fn) {
+			if (subscribers.length === 0) {
+				stop = start(set);
+			}
 
-		this.fire('state', {
-			changed,
-			previous,
-			current: this._state
-		});
-
-		this._dependents
-			.filter(dependent => {
-				const componentState = {};
-				let dirty = false;
-
-				for (let j = 0; j < dependent.props.length; j += 1) {
-					const prop = dependent.props[j];
-					if (prop in changed) {
-						componentState['$' + prop] = this._state[prop];
-						dirty = true;
-					}
-				}
+			subscribers.push(fn);
+			fn(value);
 
-				if (dirty) {
-					dependent.component._stage(componentState);
-					return true;
-				}
-			})
-			.forEach(dependent => {
-				dependent.component.set({});
-			});
-
-		this.fire('update', {
-			changed,
-			previous,
-			current: this._state
-		});
-	},
-
-	_sortComputedProperties() {
-		const computed = this._computed;
-		const sorted = this._sortedComputedProperties = [];
-		const visited = blankObject();
-		let currentKey;
-
-		function visit(key) {
-			const c = computed[key];
-
-			if (c) {
-				c.deps.forEach(dep => {
-					if (dep === currentKey) {
-						throw new Error(`Cyclical dependency detected between ${dep} <-> ${key}`);
-					}
-
-					visit(dep);
-				});
-
-				if (!visited[key]) {
-					visited[key] = true;
-					sorted.push(c);
+			return function() {
+				const index = subscribers.indexOf(fn);
+				if (index !== -1) subscribers.splice(index, 1);
+
+				if (subscribers.length === 0) {
+					stop();
```

Comment:
```suggestion
					stop && stop();
```

It seems nice to allow the `start` function to return nothing, and have that mean tearing this down is a no-op.